### DISCUSSION TITLE
无法自动更新插件

### DISCUSSION THREAD

**Discussion by uxfion** [nshen/learn-neovim-lua#146] [Q&A]
将原本plugins.lua中的自动安装移除
```
-- 每次保存 plugins.lua 自动安装插件
pcall(
    vim.cmd,
    [[
        augroup packer_user_config
        autocmd!
        autocmd BufWritePost plugins.lua source <afile> | PackerSync
        augroup end
    ]]
)
```
新增autocmds.lus
```
-- 修改lua/plugins.lua 自动更新插件
autocmd("BufWritePost", {
    group = myAutoGroup,
    -- autocmd BufWritePost plugins.lua source <afile> | PackerSync
    callback = function()
        if vim.fn.expand("<afile>") == "lua/plugins.lua" then
            vim.api.nvim_command("source lua/plugins.lua")
            vim.api.nvim_command("PackerSync")
        end
    end,
})
```
当我保存时，插件没有自动更新

**Comment by nshen** [comment]
没有`require("autocmds.lua") `吧？

**Comment by uxfion** [comment]
加了的
<img width="644" alt="image" src="https://user-images.githubusercontent.com/44778029/204135125-fe94b958-7dd3-402d-90b7-6f208af5c39e.png">
<img width="421" alt="image" src="https://user-images.githubusercontent.com/44778029/204135135-894fb967-d7b9-4413-bb6a-f38544f6ede8.png">

**Reply by uxfion** [reply]
个人配置文件如下
https://github.com/uxfion/dotfiles

**Reply by uxfion** [reply]
`autocmds.lua`中

1. 进入Terminal 自动进入插入模式
2. 保存时自动格式化
3. Highlight on yank
4. 用o换行不要延续注释

均可以使用，唯独**自动更新插件**无法使用

**Comment by jinzhongjia** [comment]
```lua
-- 修改lua/plugins.lua 自动更新插件
autocmd("BufWritePost", {
    group = myAutoGroup,
    -- autocmd BufWritePost plugins.lua source <afile> | PackerSync
    callback = function()
        if vim.fn.expand("<afile>") == "lua/plugins.lua" then
            print("进入该作用域")
            vim.api.nvim_command("source lua/plugins.lua")
            vim.api.nvim_command("PackerSync")
        end
    end,
})
```

尝试改成我上面这样然后通过`:messages`判断是否进入判断语句中，应该是没进入这个语句体

简单的lua的调试无非就是print()，然后这些信息通过`:messages`可以看到

**Reply by uxfion** [reply]
感谢，因为neovim我才刚接触的lua，已经学会最简单的lua调试。

该问题已在这里解决[Auto command does not enable automatic plugin update when plugins.lua is saved #147](https://github.com/nshen/learn-neovim-lua/issues/147)

就是把`print(vim.fn.expand("<afile>"))`打印出来，判断if是否符合条件，感谢@jinzhongjia和@nshen的指导

## 解决方法

```
-- 修改lua/plugins.lua 自动更新插件
autocmd("BufWritePost", {
    group = myAutoGroup,
    -- autocmd BufWritePost plugins.lua source <afile> | PackerSync
    callback = function()
        if vim.fn.expand("<afile>") == ".config/nvim/lua/plugins.lua" then
            vim.api.nvim_command("source .config/nvim/lua/plugins.lua")
            vim.api.nvim_command("PackerSync")
        end
    end,
})
```
将`if vim.fn.expand("<afile>") == "lua/plugins.lua" then`修改为`if vim.fn.expand("<afile>") == ".config/nvim/lua/plugins.lua" then`，就正常了

暂时还不清楚为什么大家的是`"lua/plugins.lua"`，而我的是`".config/nvim/lua/plugins.lua"`，neovim版本问题？

**Reply by nshen** [reply]
也许是个bug，vim.fn.expand("<afile>") 难道跟所在目录有关？我也不太清楚，有空研究一下。
